File: in_Java/0054 Spiral Matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> res = new ArrayList<>();
        if(matrix == null || matrix.length == 0) return res;
        int row = matrix.length;
        int col = matrix[0].length;
        int left = 0, right = col-1;
        int up = 0, down = row-1;
        while(res.size() != row * col){
            for(int j = left; j <= right && res.size() < row*col; j++){
                res.add(matrix[up][j]);
            }
            for(int j = up+1; j < down && res.size() < row*col; j++){
                res.add(matrix[j][right]);
            }
            for(int j = right; j >= left && res.size() < row*col; j--){
                res.add(matrix[down][j]);
            }
            for(int j = down-1; j > up && res.size() < row*col; j--){
                res.add(matrix[j][left]);
            }
            left++; right--;
            up++; down--;
        }
        return res;
    }
}

File: in_Python/0054 Spiral Matrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if matrix == None or len(matrix) == 0 or len(matrix[0]) == 0:
            return []
        row = len(matrix)
        col = len(matrix[0])
        up, down = 0, row-1
        left,right = 0, col-1
        res = []
        while len(res) != row * col:
            for i in range(left, right+1):
                if len(res) != row * col:
                    res.append(matrix[up][i])
            for i in range(up+1, down):
                if len(res) != row * col:
                    res.append(matrix[i][right])
            for i in range(right, left-1,-1):
                if len(res) != row * col:
                    res.append(matrix[down][i])
            for i in range(down-1, up,-1):
                if len(res) != row * col:
                    res.append(matrix[i][left])
            up += 1
            left += 1
            right -= 1
            down -= 1
        return res